### ISSUE TITLE
GJ divided by 1000?

### ISSUE THREAD

**Issue by HAEdwin** [golles/Home-Assistant-Sensor-MC66C#7]
I wonder why the first parameter, stadsverwarming energy in GJ is divided by 1000. I have to divide it by 100 to keep the value equal to the display value. I have also included the 10th parameter because this provides info about the battery. From what I understand it can run flat when querying the Kampstrup too often. Maybe later I'll do a pull request.

**Comment by coltju** [comment]
the reading is correct with me. it displays up to three digits on the meter as well. 

battery info added is a nice addition to keep in mind!

**Comment by coltju** [comment]
can you elaborate on what the battery reading is? i only get a bunch of zeroes on it. I basically just added                 data.append(int((reading[9]).decode("utf-8"))) # Battery, and the const file ofc. 

Not sure what to do though when the battery will actual run flat, dont think the provider will be very happy with that.

**Comment by golles** [comment]
I find it a bit hard to find out which meters do have a battery and which ones don't. I think I almost ran my new meters battery flat, it wasn't reporting data for a while....

To reduce battery life, I suggest updating from 30 to 3600 seconds in https://github.com/golles/Home-Assistant-Sensor-MC66C/blob/master/custom_components/mc66c/const.py#L6

**Comment by HAEdwin** [comment]
> I find it a bit hard to find out which meters do have a battery and which ones don't. I think I almost ran my new meters battery flat, it wasn't reporting data for a while....

The manual describes the possibility to connect an external power supply

**Comment by Piepaa** [comment]
Hi, same here I had to divided with 100 too, next to this I have my MC66 connected to a external power supply of 230Vac. But I am still suffering from bad readings where only the last 2 decimals are shown. So instead of 451.22  GJ the sensor shows 0.22 GJ.

**Comment by coltju** [comment]
Didnt you get an issue with the provider? iirc you have to open the meter to add an external power supply right ?

**Comment by Piepaa** [comment]
I was the first on the "blok verwarming" that installed floor heating. So a GJ-meter had to be installed and in those days it was powered by an external connection.

**Comment by HAEdwin** [comment]
Oke, so there are models where you need to divide by 100 (like the 66C02F1387) any other models where this applies? Which models have to divide by 1000?

As for the bad readings, I noticed that there are far and near infrared readingheads. I have fair results (9 out of 10 with an interval of 60 minutes) with this one: https://nl.aliexpress.com/item/1005004351799379.html but I can't guarantee it works for you. 

Unfortunately there is no other way than to ignore/drop faulty readings. I've added an extra check on datafield length in my repository might do a PR but am waiting for more input on which meters need /100 and which /1000
`2022-10-27 17:01:20.721 WARNING (SyncWorker_3) [custom_components.mc66c.sensor] Error=Received invalid datafield length parsing data=[b'c72325', b'0721210', b'0166393', b'0005675', b'0005069', b'0000606', b'0000000', b'0000007', b'0000059', b'0000000']`

**Comment by Piepaa** [comment]
I have installed a 66C03F1385 an divided by 100.
About the bad readings. I noticed that during the summer where I have a closed the main valve, there are no bad readings?
The sensor I use is; https://www.photovoltaikforum.com/thread/145886-habe-lesk%C3%B6pfe-mit-usb-%C3%BCber/

**Comment by HAEdwin** [comment]
I think that should not make a difference unless the meter has a defect, I close the valve also but opening the valve did not cause any other weird data. Any chance on false readings due to (sun)light on the meter?

**Comment by golles** [comment]
> Unfortunately there is no other way than to ignore/drop faulty readings. I've added an extra check on datafield length in my repository might do a PR but am waiting for more input on which meters need /100 and which /1000 `2022-10-27 17:01:20.721 WARNING (SyncWorker_3) [custom_components.mc66c.sensor] Error=Received invalid datafield length parsing data=[b'c72325', b'0721210', b'0166393', b'0005675', b'0005069', b'0000606', b'0000000', b'0000007', b'0000059', b'0000000']`

Feel free to create any PR's with improvements. I can also make you a collaborator in this repo

**Comment by mathijsvandenberg** [comment]
For me, i have to divide by 100 as well (Kamstrup 66C2AF1385). The value on the display is 00391.78GJ, the value in homeassistant is 39.1780 GJ

**Comment by HAEdwin** [comment]
@mathijsvandenberg @Piepaa  You are right both must be divided by 100 (on 66C02F1387 and 66C2AF1385 - so I suppose the whole 66C series)

**Comment by David-tech42** [comment]
Hello,
FYI: just installed it and I also had to divide both by 100 instead of 1000.
model: 66C03F1385
cable: https://nl.aliexpress.com/item/1005003259326667.html?spm=a2g0o.order_list.order_list_main.5.2d1279d2aweLr5&gatewayAdapt=glo2nld

still great job !

**Comment by HindrikDeelstra** [comment]
(Fairly) new HA user here, with an IR head reading out my MC66C.

I also have to divide both the "Energy" and "Volume" data by 100 (not 1000) to get the correct numbers, and I also need to divide the "Flow" data by 1 (not 10), to get the correct number.

Just FYI ;)

**Comment by golles** [comment]
I've no idea why some meters behave differently, my suggestion is to make the change locally on your installation

**Comment by HindrikDeelstra** [comment]
> I've no idea why some meters behave differently, my suggestion is to make the change locally on your installation

I already did that, by changing the relevant multipliers in the local "sensor.py" file. But that doesn't seem like a very update-friendly way to go. If we should not expect updates anytime soon, I guess it's OK to keep it this way.

Or, I could create some helpers to adjust the few values that need it, but that also seems a bit overkill, and messes with the elegant way to add sensors to cards, etc...

Oh well, I'm just waiting for a new (smart) heat meter, which will use a different integration anyway. I'll survive! ;)
